### ISSUE TITLE
6월

### ISSUE THREAD

**Issue by Seyoung9304** [Seyoung9304/Seyoung9304.github.io#5]
# Model Finetuning with KITTI dataset
- KITTI Dataset
    - real life blackbox images
    - inaccurate flow data
    - 200 ref-target image pairs with ground truth
    - 400 test image (no label)
    - 이미지별로 사이즈 다 달라서 [368, 1232]로 crop하여 학습
- Fine tuning
    - FlyingChairs [384, 512]로 학습된 small correlation residual block 모델 가중치 load하여 학습
    - generalization을 목적으로 하는 fine tuning -> weight가 KITTI dataset으로 편향되는 것을 방지하고자 normalization layer freeze (내가 언급한 이유는 뇌피셜임. https://gaussian37.github.io/dl-concept-batchnorm/, https://stackoverflow.com/questions/63016740/why-its-necessary-to-frozen-all-inner-state-of-a-batch-normalization-layer-when 참고하였음 - Batch norm freezing의 효과에 대해서는 조금 더 생각해봐야 할 것 같다.)
    
    - 학습이 제대로 되지 않음
        - Train loss가 1epoch부터 상승, 이후 다시 감소
        - Valid loss가 계속 상승 후 수렴
        - (문제 해결 시도 1) Data augmentation 추가 (random crop, scale)
            - 데이터 개수가 너무 적어서 금방 overfitting 되는 것은 아닐까?하는 생각에 시도해봄
            - 효과없음
            - Flow data라서 rotate 등의 augmentation을 구현하기가 어려웠음
            - 더 다양한 augmentation을 시도해보았으면 어땠을지 궁금함
        - (문제 해결 시도 2) Freezing extractor layer / gru unit 
            - 효과 없음
    - 추정되는 실패 원인
        - FlyingChairs로만 학습된 모델 (즉, general하지 않은 모델)에 너무 **작고, 데이터 특성이 다른** 데이터셋을 학습시키려고 했음
        - 1epoch만에 overfitting된 것으로 생각됨
        - 실제로, RAFT의 training schedule을 참고해보면 FC, FT, Sintel 등의 데이터로 먼저 학습한 뒤에 KITTI를 학습시킴. 
    - 해결방안
        - RAFT training schedule 따라서 학습 시도
        - AutoFlow 이용하여 학습
            - random한 객체들이 합성된 이미지로, 도메인에 국한받지 않고 다양한 이미지 학습 가능할 듯

# Searching Optimal Channel Size
Real life dataset inference하여 warp해본 결과, flow가 조금씩 밀리는 모습. (특정한 부분이 틀리기보다 전반적으로 조금씩 오차가 있었다)
small correlation을 구현하면서, feature extractor을 지나고 나면 image w, h 각각 1/8씩 downsizing되게 설계되어있다. 이 과정에서 channel 수는 계속 32로 고정했었다. 여기서 정보 손실이 일어난거 아닐까? 채널 수를 조정해보면 어떨까? 하는 생각에 실험 시작
- feature extractor 부분에서 downsampling이 일어나는 부분이 총 3번 있음. 각각의 out channel 수를
    - 32-32-64
    - 32-48-64
    - 16-32-64 
   로 바꾸어보면서 실험. 

## Experiment Result
* Residual block으로만 실험했음

|Channel|EPE|PSNR|MACs|
|:----:|:----:|:----:|:--:|
|32-32-32|1.6464709	|18.39967345|2522087424
|32-32-64|1.4002583	|18.51621593|4243193856
|32-48-64|1.3746147	|18.54481756|5790892032

- 시각화 결과, 경계 부분을 더 뚜렷하게 잘 잡아냄
- 그러나 여전히 flow가 조금씩 밀리는 부분은 존재
    - downsampling이 너무 과하게 일어나는 것은 아닐지. 연산량때문에 small corr을 택했는데, 다른 부분에서 연산량을 줄이고 corr은 그대로 둬야할지. 
    - extractor 외에, flow를 predict하는 gru unit에 대한 연구도 필요할 것 같다. 
- flow가 큰 부분을 잘 잡지 못하는 문제 있음 -> extractor kernel size를 늘려보는 시도를 해보고 싶음
- naive하게 채널 수를 늘린거라, 분명이 redundant한 연산이 있을 것임. layer별 채널 크기와 성능 사이 연관성 파악하여, critical factor만 남기고 불필요한 연산을 줄일 필요 있음. 

# 회고
- 실험 결과가 왜 이렇게 나왔을지 생각하고, 증명하고, 그것을 토대로 다음 step을 정하는 과정에 익숙해질 필요가 있음. 지금은 살짝 무지성 느낌
- 타인에게 실험의 당위성/결과/next step을 매끄럽게 논리적으로 정리해서 말하는 연습할 필요가 있음
- 발표자료를 매끄럽게 준비하는 연습. 처음 보는 사람들도 쉽게 이해 가능하도록. 
- 실험에 대한 insight는 대체 어디서 얻을 수가 있을까?
- idea를 바로 실행할 수 있는 힘을 기른 것 같아서 굿
- 조금 더 적극적으로 소통하는 연습하기
- 고생많았다 많이배웠다 최고!

**Comment by Seyoung9304** [comment]
Issues
# Train with multi-gpu

## command
```
# command
$ CUDA_AVAILABLE_DEVICES=0,1      (device numbers connected with comma)
# in __main__()
os.environ["CUDA_DEVICE_ORDER"]="PCI_BUS_ID"
```

## model
### Loading model
모델은 [nn.DataParallel](https://pytorch.org/docs/stable/generated/torch.nn.DataParallel.html?highlight=dataparallel#torch.nn.DataParallel) 이용해서 wrap해준다. 
이때, 로드할 ckpt가 있다면 `nn.DataParallel`으로 wrap하기 전에 load해주어야 한다. 
```
_model = MyModel()
if (args.ckpt is not None):
        checkpoint = torch.load(args.ckpt)
        _model.load_state_dict(checkpoint['model'], strict=False)
model = nn.DataParallel(_model).to(device)
```
위 코드 처럼 해야함 (모델 instance 생성 -> ckpt load -> dataparallel)
ckpt를 나중에 로딩했더니 그냥 깡통모델이 불러와졌다. 디버깅 하는데 고생 좀 했다 ㅜㅜ

### Saving model
`nn.DataParallel`을 이용한 경우, `model.state_dict()`로 가중치를 불러올 수 없다. `model.module.state_dict()`로 접근해야 한다고 함.
그래서 모델 저장 시, 
```
torch.save({'model': model.module.state_dict()}}
```
로 해야 함. 
multi gpu이든 아니든 `.module`로 접근하는게 더 general하다고 한다. 

https://pytorch.org/tutorials/beginner/saving_loading_models.html#saving-torch-nn-dataparallel-models

> 더 많은 gpu, dataloader worker를 사용한다고 무조건 좋은 것은 아니다. cpu에서 처리하는 연산량을 고려하지 않으면 병목현상이 발생할 수 있음. 짬날때 관련해서 공부좀 해야겠다. 

## Freezing
모델 클래스 내부에 `freeze_bn`, `freeze_extractor` 이런식으로 함수를 구현해두고 사용했음. 
### 특정 연산을 모두 freeze하고 싶은 경우

```
for m in self.modules():
    if isinstance(m, (nn.BatchNorm2d, nn.InstanceNorm2d, nn.GroupNorm)):
        m.eval()
```

### nn.Sequential로 묶여있는 한 단위를 freeze하고 싶은 경우

sequential하게 묶여있으면, 묶여있는 연산들이 하나의 모듈로 처리가 된다. 그래서 Sequential로 묶여있는 애들 하나하나를 보고싶으면 또 그 하위 단위로 내려가야 함. 
아래 코드를 보면 for 문이 두번이다. 

```
for m in self.parameters():
    m.requires_grad = False
for name, module in self.named_modules():
    if name.split(".")[0] in ['motion_encoder', 'gru', 'flow_head', 'mask']:
        for n, param in module.named_parameters():
            param.requires_grad = True
```


이 과정에서 `.eval()`과 `requires_grad = False`의 차이점이 궁금했다. 

**주의점**
validation phase 시, 보통
```
model.eval()
with torch.no_grad():
  # Evaluate model
  # ...
model.train()
```
위와 같은 구조를 가지게 된다. 
이때, model.train()시 내부 모든 파라미터가 trainable해지기 때문에, 다시 freeze 해줘야 한다. 
```
model.eval()
with torch.no_grad():
  # Evaluate model
  # ...
model.train()
model.module.freeze_bn()
```

이런식으로~
